### PR TITLE
Fixed equip ranged weapons (Guns, throwns etc..)

### PR DESCRIPTION
**Pull request by ghost**
Now characters are able to equip this types of item

### CODE REVIEW

**Comment by fallcons** [comment]
Just to know, VirusEntertainment, are you working in the weapon of the hunter ?
thx

**Comment by VirusEntertainment** [comment]
atm i'm working on new locales ^^ https://github.com/gegge6265/TrinityCore/commits/locales

**Comment by fallcons** [comment]
Haha ok thx ;)
btw good work !

**Comment by VirusEntertainment** [comment]
why this isn't merged?

**Comment by tkrokli** [comment]
I don't know, but I can guess that one reason could be that this PR does not have a description that it "closes issue #14652" for instance. A good and formal description, as well as linking it to open issues, is a sign of active work to make the source better. If you are in doubt, open the last 50 commits and look for the variations of header/description text, that should give you a clue.

**Comment by martial69320** [comment]
I have apply your patch, i can equip ranged weapons, but now, hunter do 0 damage with spells and auto shot.

File: src/server/game/Entities/Item/ItemTemplate.h [hunk 1]
```diff
@@ -697,6 +697,13 @@ struct ItemTemplate
                GetSubClass() == ITEM_SUBCLASS_WEAPON_GUN ||
                GetSubClass() == ITEM_SUBCLASS_WEAPON_CROSSBOW;
     }
+    
+    bool IsRangedInventoryType() const
+    {
+        return InventoryType == INVTYPE_RANGED ||
++               InventoryType == INVTYPE_THROWN ||
++               InventoryType == INVTYPE_RANGEDRIGHT;
```
Comment (DDuarte) [inline src/server/game/Entities/Item/ItemTemplate.h]: Extra + here, won't compile

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -9606,7 +9606,27 @@ uint8 Player::FindEquipSlot(ItemTemplate const* proto, uint32 slot, bool swap) c
...
         case INVTYPE_RANGED:
-            slots[0] = EQUIPMENT_SLOT_RANGED;
+            slots[0] = EQUIPMENT_SLOT_MAINHAND;
+            if (Item* mhWeapon = GetItemByPos(INVENTORY_SLOT_BAG_0, EQUIPMENT_SLOT_MAINHAND))
+            {
+                if (ItemTemplate const* mhWeaponProto = mhWeapon->GetTemplate())
+                {
+                    if (mhWeaponProto->GetSubClass() == ITEM_SUBCLASS_WEAPON_POLEARM || mhWeaponProto->GetSubClass() == ITEM_SUBCLASS_WEAPON_STAFF)
+                    {
+                        const_cast<Player*>(this)->AutoUnequipOffhandIfNeed(true);
```
Comment (DDuarte) [inline src/server/game/Entities/Player/Player.cpp]: These const_casts do not look good. You are unequipping a weapon in a method named "FindEquipSlot". If it's a "find" it shouldn't change anything (a.k.a not const).

File: src/server/game/Entities/Item/ItemTemplate.h [hunk 3]
```diff
@@ -697,6 +697,13 @@ struct ItemTemplate
                GetSubClass() == ITEM_SUBCLASS_WEAPON_GUN ||
                GetSubClass() == ITEM_SUBCLASS_WEAPON_CROSSBOW;
     }
+    
+    bool IsRangedInventoryType() const
```
Comment (DDuarte) [inline src/server/game/Entities/Item/ItemTemplate.h]: This is not being used anywhere.

File: src/server/game/Entities/Item/ItemTemplate.h [hunk 4]
```diff
@@ -697,6 +697,13 @@ struct ItemTemplate
...
+    {
+        return InventoryType == INVTYPE_RANGED ||
```
Comment (cemak) [inline src/server/game/Entities/Item/ItemTemplate.h]: Rename InventoryType to GetInventoryType()

**Comment by spuresz** [comment]
yes hunter bow invisible

**Comment by spuresz** [comment]
can see the bow? just 0 damage?

**Comment by fallcons** [comment]
Can`t see the bow on the last realese 6.2

File: src/server/game/Entities/Item/ItemTemplate.h (hunk 4 above)
Comment (tkrokli) [inline src/server/game/Entities/Item/ItemTemplate.h]: Another detail which makes the travis build error confusing, referring line 764, is that @VirusEntertainment is using @gegge's repo from 11. May this year. gegge's file is only 716 lines long, while the TC repo file has 771 lines.

File: src/server/game/Entities/Item/ItemTemplate.h (hunk 1 above)
Comment (Aokromes) [inline src/server/game/Entities/Item/ItemTemplate.h]: 4 months and still unfixed.

**Comment by tkrokli** [comment]
@VirusEntertainment : check out https://github.com/xIchigox/6.x/commit/b622499c32750244c00b061cebb24274c41557ad for a hint on what will work for [Bow], at least.
BTW, @gegge's fork is 1103 commits behind TrinityCore:6.x (2015-05-11), either update it or close this PR.

**Comment by Keldo** [comment]
using trokli's idff patch I was able to get Hunter to equip and use on 6x
https://gist.github.com/tkrokli/eb5c0e3b5507b78ff279

**Comment by tkrokli** [comment]
Notice that my diff gist is only copied code from @xIchigox (https://github.com/xIchigox/6.x) :
https://github.com/xIchigox/6.x/blob/master/src/server/game/Entities/Player/Player.cpp#L9699-L9719
and
https://github.com/xIchigox/6.x/blob/master/src/server/game/Entities/Item/ItemTemplate.h#L762-L768
(notice that he has modified his ItemTemplate.h compared to what I used in my gist)
I could of course update my gist, but there are already 2 main TC devs disliking that code.

**Comment by spuresz** [comment]
Good! :)

**Comment by spuresz** [comment]
Soon will added?

**Comment by Shauren** [comment]
This barely scratches the surface of changes needed to make ranged weapons go to mainhand slot.

**Pull request closed without merging**